pytorch_lightning/loops/optimization/manual_loop.py

```diff
@@ -11,16 +11,89 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from typing import Any, Optional
+from dataclasses import dataclass, field
+from typing import Any, Dict, Optional
+
+from torch import Tensor
 
 from pytorch_lightning.loops import Loop
-from pytorch_lightning.loops.optimization.closure import ClosureResult
+from pytorch_lightning.loops.optimization.closure import OutputResult
 from pytorch_lightning.loops.utilities import (
     _build_training_step_kwargs,
     _check_training_step_output,
     _extract_hiddens,
     check_finite_loss,
 )
+from pytorch_lightning.utilities.apply_func import apply_to_collection
+from pytorch_lightning.utilities.memory import recursive_detach
+from pytorch_lightning.utilities.types import STEP_OUTPUT
+from pytorch_lightning.utilities.warnings import rank_zero_deprecation
+
+
+@dataclass
+class ManualResult(OutputResult):
+    """A container to hold the result returned by the ``ManualLoop``.
+
+    It is created from the output of :meth:`~pytorch_lightning.core.lightning.LightningModule.training_step`.
+
+    Attributes:
+        closure_loss: The loss with a graph attached.
+        loss: A detached copy of the closure loss.
+        extra: Any keys other than the loss returned.
+    """
+
+    closure_loss: Optional[Tensor]
+    loss: Optional[Tensor] = field(init=False, default=None)
+    extra: Dict[str, Tensor] = field(default_factory=dict)
+
+    def __post_init__(self) -> None:
+        # TODO: remove with the deprecation removal in v1.6
```

Comment:
I am not sure about the `deprecation` there. I believe users won't detach properly the extras as it wasn't done in the past.

Reply:
Changing this would be out of the scope of this PR.

The plan is to keep the `grad_fn` check but not `detach` it automatically, instead fail.
This was chosen to avoid silent errors where users don't know their outputs have a `grad_fn`.